src/rdkafka_request.c

```diff
@@ -3294,16 +3294,21 @@ rd_kafka_handle_Produce_parse(rd_kafka_broker_t *rkb,
         const int log_decode_errors = LOG_ERR;
         int64_t log_start_offset    = -1;
 
-        rd_kafka_buf_read_i32(rkbuf, &TopicArrayCnt);
+        rd_kafka_buf_read_arraycnt(rkbuf, &TopicArrayCnt, RD_KAFKAP_TOPICS_MAX);
         if (TopicArrayCnt != 1)
                 goto err;
 
         /* Since we only produce to one single topic+partition in each
          * request we assume that the reply only contains one topic+partition
          * and that it is the same that we requested.
          * If not the broker is buggy. */
-        rd_kafka_buf_skip_str(rkbuf);
-        rd_kafka_buf_read_i32(rkbuf, &PartitionArrayCnt);
+        rd_kafkap_str_t topic_name;
+        if (request->rkbuf_reqhdr.ApiVersion >= 9)
+                rd_kafka_buf_read_str(rkbuf, &topic_name);
+        else
+                rd_kafka_buf_skip_str(rkbuf);
```

Comment:
You can make `rd_kafka_buf_skip_str` generic and use it here, while you can use `rd_kafka_buf_skip_str_no_flexver` in mock broker header parsing as flexver isn't allowed there in any case.

```c
/**
 * Skip a string.
 */
#define rd_kafka_buf_skip_str_no_flexver(rkbuf)                                \
        do {                                                                   \
                int16_t _slen;                                                 \
                rd_kafka_buf_read_i16(rkbuf, &_slen);                          \
                rd_kafka_buf_skip(rkbuf, RD_KAFKAP_STR_LEN0(_slen));           \
        } while (0)

/**
 * Skip a string.
 */
#define rd_kafka_buf_skip_str(rkbuf)                                           \
        do {                                                                   \
                if ((rkbuf)->rkbuf_flags & RD_KAFKA_OP_F_FLEXVER) {            \
                        uint64_t _uva;                                         \
                        rd_kafka_buf_read_uvarint(rkbuf, &_uva);               \
                        rd_kafka_buf_skip(                                     \
                            rkbuf, RD_KAFKAP_STR_LEN0(((int64_t)_uva) - 1));   \
                } else {                                                       \
                        rd_kafka_buf_skip_str_no_flexver(rkbuf);               \
                }                                                              \
        } while (0)
```

Reply:
Updated